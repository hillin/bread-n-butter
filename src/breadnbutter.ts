interface Parser<Value> {
  (context: ParserContext): Result<Value>;
}

interface ParserContext {
  type: "bnb.ParserContext";
  input: string;
  index: number;
}

interface OK<Value> {
  type: "bnb.OK";
  value: Value;
  context: ParserContext;
}

interface Fail {
  type: "bnb.Fail";
  message: string;
  context: ParserContext;
}

type Result<Value> = OK<Value> | Fail;

export function custom<Value>(
  handler: (context: ParserContext) => Result<Value>
): Parser<Value> {
  return handler;
}

export function ok<Value>(value: Value, context: ParserContext): OK<Value> {
  return {
    type: "bnb.OK",
    value,
    context,
  };
}

export function fail(message: string, context: ParserContext): Fail {
  return {
    type: "bnb.Fail",
    message,
    context,
  };
}

export function parse<Value>(
  parser: Parser<Value>,
  input: string
): Result<Value> {
  const context: ParserContext = {
    type: "bnb.ParserContext",
    input,
    index: 0,
  };
  const result = andThen(parser, matchEOF)(context);
  if (result.type === "bnb.OK") {
    const [value] = result.value;
    return ok(value, result.context);
  }
  return result;
}

export function of<Value>(value: Value): Parser<Value> {
  return custom((context) => ok(value, context));
}

export function andThen<Value1, Value2>(
  parser1: Parser<Value1>,
  parser2: Parser<Value2>
): Parser<readonly [Value1, Value2]> {
  return custom((context) => {
    const result1 = parser1(context);
    if (result1.type === "bnb.Fail") {
      return result1;
    }
    const result2 = parser2(result1.context);
    if (result2.type === "bnb.Fail") {
      return result2;
    }
    const value = [result1.value, result2.value] as const;
    return ok(value, result2.context);
  });
}

export function matchString(string: string): Parser<string> {
  return custom((context) => {
    const chunk = context.input.slice(
      context.index,
      context.index + string.length
    );
    if (chunk === string) {
      return ok(string, {
        ...context,
        index: context.index + 1,
      });
    }
    return fail(string, context);
  });
}

export const matchEOF = custom((context) => {
  if (context.index === context.input.length) {
    return ok("end of file", context);
  }
  return fail("end of file", context);
});

export function matchRegExp(regexp: RegExp): Parser<string> {
  return custom((context) => {
    const stickyRegexp = new RegExp(regexp, "y");
    stickyRegexp.lastIndex = context.index;
    const match = context.input.match(stickyRegexp);
    console.log("MATCH", match);
    if (!match) {
      return fail(regexp.toString(), context);
    }
    return ok(match[0], {
      ...context,
      index: context.index + match[0].length,
    });
  });
}
