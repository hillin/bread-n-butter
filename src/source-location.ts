/**
 * Represents a location in the input (source code). Keeps track of `index` (for
 * use with `.slice` and such), as well as `line` and `column` for displaying to
 * users.
 */
export class SourceLocation {
  /** The string index into the input (e.g. for use with `.slice`) */
  index: number;
  /**
   * The line number for error reporting. Only the character `\n` is used to
   * signify the beginning of a new line.
   */
  line: number;
  /**
   * The column number for error reporting.
   *
   * Note that certain complex emojis may count as more than one column, because
   * they're actually several emojis in a trenchcoat (e.g. family emojis.)
   *
   * The column number is incremented based on each character as defined by
   * ES2015's string iteration protocol (i.e. the result of calling
   * `Array.from(string)`).
   */
  column: number;

  /** @ignore */
  constructor(index: number, line: number, column: number) {
    this.index = index;
    this.line = line;
    this.column = column;
  }

  /** @ignore */
  add(chunk: string): SourceLocation {
    let { index, line, column } = this;
    for (const ch of chunk) {
      index++;
      if (ch === "\n") {
        line++;
        column = 1;
      } else {
        column++;
      }
    }
    return new SourceLocation(index, line, column);
  }
}
