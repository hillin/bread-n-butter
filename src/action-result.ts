import { SourceLocation } from "./source-location";

export type ActionResult<A> = ActionOK<A> | ActionFail;

export class ActionOK<A> {
  location: SourceLocation;
  value: A;
  furthest: SourceLocation;
  expected: string[];

  constructor(
    location: SourceLocation,
    value: A,
    furthest: SourceLocation,
    expected: string[]
  ) {
    this.location = location;
    this.value = value;
    this.furthest = furthest;
    this.expected = expected;
  }

  isOK(): this is ActionOK<A> {
    return true;
  }

  merge<B>(b: ActionResult<B>): ActionResult<B> {
    return merge(b, this);
  }
}

export class ActionFail {
  furthest: SourceLocation;
  expected: string[];

  constructor(furthest: SourceLocation, expected: string[]) {
    this.furthest = furthest;
    this.expected = expected;
  }

  isOK<A>(): this is ActionOK<A> {
    return false;
  }

  merge<B>(b: ActionResult<B>): ActionResult<B> {
    return merge(b, this);
  }
}

function merge<A, B>(a: ActionResult<A>, b: ActionResult<B>): ActionResult<A> {
  if (a.furthest.index > b.furthest.index) {
    return a;
  }
  const expected =
    a.furthest.index === b.furthest.index
      ? union(a.expected, b.expected)
      : b.expected;
  if (a.isOK()) {
    return new ActionOK(a.location, a.value, b.furthest, expected);
  }
  return new ActionFail(b.furthest, expected);
}

function union(a: string[], b: string[]): string[] {
  return [...new Set([...a, ...b])].sort();
}
