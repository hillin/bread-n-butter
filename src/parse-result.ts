import { SourceLocation } from "./source-location";

export type ParseResult<A> = ParseOK<A> | ParseFail;

export class ParseOK<A> {
  readonly value: A;

  constructor(value: A) {
    this.value = value;
  }

  isOK(): this is ParseOK<A> {
    return true;
  }
}

export class ParseFail {
  // TODO: Possibly yoink the properties off location and put them straight on
  // this class so it's easier to look at the object?
  readonly location: SourceLocation;
  readonly expected: readonly string[];

  constructor(location: SourceLocation, expected: readonly string[]) {
    this.location = location;
    this.expected = expected;
  }

  isOK<A>(): this is ParseOK<A> {
    return false;
  }
}
