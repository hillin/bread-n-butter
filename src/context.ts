import { SourceLocation } from "./source-location";
import { ActionResult, ActionOK, ActionFail } from "./action-result";

export class Context {
  readonly input: string;
  readonly location: SourceLocation;

  constructor(input: string, location: SourceLocation) {
    this.input = input;
    this.location = location;
  }

  withLocation(location: SourceLocation): Context {
    return new Context(this.input, location);
  }

  move(index: number): SourceLocation {
    const start = this.location.index;
    const end = index;
    const chunk = this.input.slice(start, end);
    return this.location.add(chunk);
  }

  ok<A>(index: number, value: A): ActionResult<A> {
    return new ActionOK(
      this.move(index),
      value,
      new SourceLocation(-1, -1, -1),
      []
    );
  }

  fail<A>(index: number, expected: readonly string[]): ActionResult<A> {
    return new ActionFail(this.move(index), expected);
  }
}
