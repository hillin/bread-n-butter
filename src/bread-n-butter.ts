import { ActionResult, ActionFail } from "./action-result";
import { Context } from "./context";
import { ParseFail, ParseOK, ParseResult } from "./parse-result";
import { SourceLocation } from "./source-location";

export class Parser<A> {
  readonly action: (context: Context) => ActionResult<A>;

  /**
   * Creates a new custom parser that performs the given parsing action.
   * **Note:** That use of this constructor is an advanced feature and not
   * needed for most parsers.
   *
   * ```ts
   * import * as bnb from "bread-n-butter";
   *
   * const number = new bnb.Parser((context) => {
   *   const start = context.location.index;
   *   const end = context.location.index + 2;
   *   if (context.input.slice(start, end) === "AA") {
   *     // Return how far we got, and what value we found
   *     return context.ok(end, "AA");
   *   }
   *   // Return how far we got, and what value we were looking for
   *   return context.fail(start, "AA");
   * });
   * ```
   */
  constructor(action: (context: Context) => ActionResult<A>) {
    this.action = action;
  }

  /**
   * Returns a [[ParseResult]] with the parse value if successful, otherwise a
   * failure value indicating where the error is and what values we were looking
   * for at the time of failure. Use [[isOK]] to check if the parse succeeded or
   * not.
   *
   * ```ts
   * import * as bnb from "bread-n-butter";
   *
   * const a = bnb.str("a");
   * const result1 = a.parse("a");
   * if (result.isOK()) {
   *   console.log(result.value);
   *   // => "a"
   * } else {
   *   const { location, expected } = result;
   *   console.error("error at line", location.line, "column", location.column);
   *   console.error("expected one of", expected.join(", "));
   * }
   * ```
   */
  parse(input: string): ParseResult<A> {
    const location = new SourceLocation(0, 1, 1);
    const context = new Context(input, location);
    const result = this.and(eof).action(context);
    if (result.isOK()) {
      return new ParseOK(result.value[0]);
    }
    return new ParseFail(result.furthest, result.expected);
  }

  /**
   * Combines two parsers one after the other, yielding the results of both in
   * an array.
   *
   * ```ts
   * import * as bnb from "bread-n-butter";
   *
   * const a = bnb.str("a");
   * const b = bnb.str("b");
   * const ab = a.and(b);
   * const result = ab.parse("ab");
   * console.log(result.value);
   * // => ["a", "b"]
   * ```
   */
  and<B>(parserB: Parser<B>): Parser<readonly [A, B]> {
    return new Parser((context) => {
      const a = this.action(context);
      if (!a.isOK()) {
        return a;
      }
      const b = a.merge(parserB.action(context.withLocation(a.location)));
      if (b.isOK()) {
        const value = [a.value, b.value] as const;
        return b.merge(context.ok(b.location.index, value));
      }
      return b;
    });
  }

  /**
   * Try to parse using the current parser. If that fails, parse using the
   * second parser.
   *
   * This is good for parsing things like _expressions_ or
   * _statements_ in programming languages, where many different types of things
   * are applicable.
   *
   * ```ts
   * import * as bnb from "bread-n-butter";
   *
   * const a = bnb.str("a");
   * const b = bnb.str("b");
   * const ab = a.or(b);
   *
   * console.log(ab.parse("a").value);
   * // => "a"
   *
   * console.log(ab.parse("b").value);
   * // => "b"
   * ```
   *
   * **Tip:** You can also use this to implement optional parsers:
   *
   * ```ts
   * import * as bnb from "bread-n-butter";
   *
   * const aMaybe = bnb.str("a").or(bnb.ok(null));
   *
   * console.log(aMaybe.parse("a").value);
   * // => "a"
   *
   * console.log(aMaybe.parse("").value);
   * // => null
   * ```
   */
  or<B>(parserB: Parser<B>): Parser<A | B> {
    return new Parser<A | B>((context) => {
      const a = this.action(context);
      if (a.isOK()) {
        return a;
      }
      return a.merge(parserB.action(context));
    });
  }

  chain<B>(fn: (value: A) => Parser<B>): Parser<B> {
    return new Parser((context) => {
      const a = this.action(context);
      if (!a.isOK()) {
        return a;
      }
      const parserB = fn(a.value);
      return a.merge(parserB.action(context.withLocation(a.location)));
    });
  }

  map<B>(fn: (value: A) => B): Parser<B> {
    return this.chain((a) => {
      return ok(fn(a));
    });
  }

  thru<B>(fn: (parser: this) => B): B {
    return fn(this);
  }

  desc(name: string): Parser<A> {
    return new Parser((context) => {
      const result = this.action(context);
      if (result.isOK()) {
        return result;
      }
      return new ActionFail(result.furthest, [name]);
    });
  }

  wrap<B, C>(before: Parser<B>, after: Parser<C>): Parser<A> {
    return before.and(this).chain(([, value]) => {
      return after.map(() => value);
    });
  }

  trim<B>(beforeAndAfter: Parser<B>): Parser<A> {
    return this.wrap(beforeAndAfter, beforeAndAfter);
  }

  many0(): Parser<readonly A[]> {
    return this.many1().or(ok([]));
  }

  many1(): Parser<readonly A[]> {
    return new Parser((context) => {
      const items: A[] = [];
      let result = this.action(context);
      while (result.isOK()) {
        items.push(result.value);
        context = context.withLocation(result.location);
        result = result.merge(this.action(context));
      }
      return result.merge(context.ok(context.location.index, items));
    });
  }

  separatedBy0<B>(separator: Parser<B>): Parser<readonly A[]> {
    return this.separatedBy1(separator).or(ok([]));
  }

  separatedBy1<B>(separator: Parser<B>): Parser<readonly A[]> {
    return this.chain((first) => {
      return separator
        .and(this)
        .map(([, value]) => value)
        .many0()
        .map((rest) => {
          return [first, ...rest] as const;
        });
    });
  }

  node<S extends string>(name: S): Parser<ParseNode<S, A>> {
    return location.and(this).chain(([start, value]) => {
      return location.map((end) => {
        const type = "ParseNode";
        return { type, name, value, start, end } as const;
      });
    });
  }
}

export interface ParseNode<S extends string, A> {
  type: "ParseNode";
  name: S;
  value: A;
  start: SourceLocation;
  end: SourceLocation;
}

// https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/parsimmon/index.d.ts#L308
// Thanks to the DefinitelyTyped folks for this type magic here
export type Rules<Spec> = {
  [P in keyof Spec]: (lang: Language<Spec>) => Parser<Spec[P]>;
};

export type Language<Spec> = {
  [P in keyof Spec]: Parser<Spec[P]>;
};

export function language<Spec>(rules: Rules<Spec>): Language<Spec> {
  const lang = {} as Language<Spec>;
  for (const key of Object.keys(rules)) {
    const k = key as keyof Spec;
    const f = () => rules[k](lang);
    lang[k] = lazy(f);
  }
  return lang;
}

export const location = new Parser<SourceLocation>((context) => {
  return context.ok(context.location.index, context.location);
});

export function ok<A>(value: A): Parser<A> {
  return new Parser((context) => {
    return context.ok(context.location.index, value);
  });
}

export function fail<A>(expected: readonly string[]): Parser<A> {
  return new Parser((context) => {
    return context.fail(context.location.index, expected);
  });
}

export const eof = new Parser<"<EOF>">((context) => {
  if (context.location.index < context.input.length) {
    return context.fail(context.location.index, ["<EOF>"]);
  }
  return context.ok(context.location.index, "<EOF>");
});

export function str<A extends string>(string: A): Parser<A> {
  return new Parser<A>((context) => {
    const start = context.location.index;
    const end = start + string.length;
    if (context.input.slice(start, end) === string) {
      return context.ok(end, string);
    }
    return context.fail(start, [string]);
  });
}

export function match(regexp: RegExp): Parser<string> {
  return new Parser((context) => {
    const flags = regexp.ignoreCase ? "iy" : "y";
    const sticky = new RegExp(regexp.source, flags);
    const start = context.location.index;
    sticky.lastIndex = start;
    const match = context.input.match(sticky);
    if (match) {
      const end = start + match[0].length;
      const string = context.input.slice(start, end);
      return context.ok(end, string);
    }
    return context.fail(start, [String(regexp)]);
  });
}

export function lazy<A>(fn: () => Parser<A>): Parser<A> {
  let action: ((context: Context) => ActionResult<A>) | undefined;
  return new Parser(function (context) {
    if (!action) {
      action = fn().action;
    }
    return action(context);
  });
}
