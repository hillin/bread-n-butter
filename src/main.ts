import * as bnb from "./bread-n-butter";

const a = bnb.str("a");
const b = bnb.str("b");
const ab = a.and(b);
const aOrB = a.or(b);
const abOrA = ab.or(a);
const balance = bnb
  .str("(")
  .or(bnb.str("["))
  .chain((value) => {
    if (value === "(") {
      return bnb.str(")");
    } else {
      return bnb.str("]");
    }
  });
const justX = bnb.of("X");
// TODO: Are these answers correct?
console.log(aOrB.parse("a"));
console.log(aOrB.parse("b"));
console.log(aOrB.parse("ab"));
console.log(aOrB.parse(""));
console.log(abOrA.parse("ab"));
console.log(abOrA.parse("aa"));
console.log(abOrA.parse("a"));
console.log(abOrA.parse(""));
console.log(abOrA.parse("aba"));
console.log(balance.parse("()"));
console.log(balance.parse("[]"));
console.log(balance.parse("[)"));
console.log(balance.parse("["));
console.log(justX.parse(""));
