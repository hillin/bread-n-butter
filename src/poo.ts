import * as bnb from "./bread-n-butter";
