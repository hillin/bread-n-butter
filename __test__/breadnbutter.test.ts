import * as bnb from "../src/breadnbutter";

test("test a test", () => {
  const isX = bnb.matchString("x");
  expect(bnb.parse(isX, "x")).toMatchSnapshot();
});

test("andThen success", () => {
  const isX = bnb.matchString("x");
  const isY = bnb.matchString("y");
  const isXY = bnb.andThen(isX, isY);
  expect(bnb.parse(isXY, "xy")).toMatchSnapshot();
});

test("andThen failure 1", () => {
  const isX = bnb.matchString("x");
  const isY = bnb.matchString("y");
  const isXY = bnb.andThen(isX, isY);
  expect(bnb.parse(isXY, "x")).toMatchSnapshot();
});

test("andThen failure 2", () => {
  const isX = bnb.matchString("x");
  const isY = bnb.matchString("y");
  const isXY = bnb.andThen(isX, isY);
  expect(bnb.parse(isXY, "y")).toMatchSnapshot();
});

test("andThen failure 3", () => {
  const isX = bnb.matchString("x");
  const isY = bnb.matchString("y");
  const isXY = bnb.andThen(isX, isY);
  expect(bnb.parse(isXY, "yx")).toMatchSnapshot();
});

test("andThen failure 4", () => {
  const isX = bnb.matchString("x");
  const isY = bnb.matchString("y");
  const isXY = bnb.andThen(isX, isY);
  expect(bnb.parse(isXY, "")).toMatchSnapshot();
});

test("matchEOF success", () => {
  const matchEOF = bnb.matchEOF;
  expect(bnb.parse(matchEOF, "")).toMatchSnapshot();
});

test("matchEOF failure", () => {
  const matchEOF = bnb.matchEOF;
  expect(bnb.parse(matchEOF, "x")).toMatchSnapshot();
});

test("matchRegExp success", () => {
  const matchNumber = bnb.matchRegExp(/\d+/);
  expect(bnb.parse(matchNumber, "12")).toMatchSnapshot();
});

test("matchRegExp failure", () => {
  const matchNumber = bnb.matchRegExp(/\d+/);
  expect(bnb.parse(matchNumber, "abc")).toMatchSnapshot();
});
