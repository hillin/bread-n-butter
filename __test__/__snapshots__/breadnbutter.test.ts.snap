// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`andThen failure 1 1`] = `
Object {
  "context": Object {
    "index": 1,
    "input": "x",
    "type": "bnb.ParserContext",
  },
  "message": "y",
  "type": "bnb.Fail",
}
`;

exports[`andThen failure 2 1`] = `
Object {
  "context": Object {
    "index": 0,
    "input": "y",
    "type": "bnb.ParserContext",
  },
  "message": "x",
  "type": "bnb.Fail",
}
`;

exports[`andThen failure 3 1`] = `
Object {
  "context": Object {
    "index": 0,
    "input": "yx",
    "type": "bnb.ParserContext",
  },
  "message": "x",
  "type": "bnb.Fail",
}
`;

exports[`andThen failure 4 1`] = `
Object {
  "context": Object {
    "index": 0,
    "input": "",
    "type": "bnb.ParserContext",
  },
  "message": "x",
  "type": "bnb.Fail",
}
`;

exports[`andThen success 1`] = `
Object {
  "context": Object {
    "index": 2,
    "input": "xy",
    "type": "bnb.ParserContext",
  },
  "type": "bnb.OK",
  "value": Array [
    "x",
    "y",
  ],
}
`;

exports[`matchEOF failure 1`] = `
Object {
  "context": Object {
    "index": 0,
    "input": "x",
    "type": "bnb.ParserContext",
  },
  "message": "end of file",
  "type": "bnb.Fail",
}
`;

exports[`matchEOF success 1`] = `
Object {
  "context": Object {
    "index": 0,
    "input": "",
    "type": "bnb.ParserContext",
  },
  "type": "bnb.OK",
  "value": "end of file",
}
`;

exports[`test a test 1`] = `
Object {
  "context": Object {
    "index": 1,
    "input": "x",
    "type": "bnb.ParserContext",
  },
  "type": "bnb.OK",
  "value": "x",
}
`;
